_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,228 @@
+
+```scala
+def getSitemaps(data: RDD[(String, Iterable[String])]): RDD[(String, Elem)] = {
+  data.map {
+    case (date: String, all: Iterable[String]) =>
+      val groups = all.iterator.grouped(50000).zipWithIndex
+      groups.map {
+        case (ids: Seq[String], index: Int) =>
+          val indexLabel = if (index != 0 || groups.hasNext) s"-${index + 1}" else ""
+          val dateLabel = s"-${date}"
+          val sitemapKey = s"sitemap-shows${dateLabel}${indexLabel}"
+          val sitemapXml = <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">{
+            ids.map(getSitemapEntry)
+          }</urlset>
+          (sitemapKey, sitemapXml)
+      }.toArray
+  }.flatMap(x => x)
+}
+```
+
+#### Writing Sitemap Files
+
+We write a file for each sitemap. The production output goes to S3.
+
+```scala
+def writeSitemaps(sitemapsXml: RDD[(String, Elem)]) = {
+  sitemapsXml.foreach {
+    case (key: String, xml: Elem) =>
+      writeFile(xml, s"${key}.xml")
+  }
+}
+```
+
+#### Writing Sitemap Index
+
+An index is a list of all the sitemaps, also written to S3.
+
+```scala
+def getSitemapsIndex(sitemapsXml: RDD[(String, Elem)]): Elem = {
+  <sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">{
+    sitemapsXml.collect().map {
+      case (key: String, _) =>
+        <sitemap><loc>{ s"https://www.artsy.net/${key}.xml" }</loc></sitemap>
+    }
+  }</sitemapindex>
+}
+
+def writeSitemapsIndex(sitemapsXml: RDD[(String, Elem)]) = {
+  val sitemapsIndexXml = getSitemapsIndex(sitemapsXml)
+  writeFile(sitemapsIndexXml, s"sitemap-shows.xml")
+}
+```
+
+You also need to do a bit more work to purge files that are no longer generated as data changes. For the shows example above we fetch a list of sitemap files such prefixed by `sitemap-shows-` and delete the set difference.
+
+```scala
+def purgeSitemaps(sitemapsXml: RDD[(String, Elem)]) = {
+  val sitemapKeys = sitemapsXml.keys.map(k => s"${k}.xml")
+  val existingFiles = ... // fetch files prefixed by sitemap-shows-
+  val filesToDelete = existingFiles &~ sitemapKeys.collect().toSet
+  keysToDelete.foreach(file => deleteFile(file))
+}
+```
+
+#### Serving Sitemaps
+
+We serve sitemaps through a Node.js proxy [in our front-end](https://github.com/artsy/force/blob/9eedf063d024ea94e6c32b01497c3fcb7f596d68/desktop/apps/sitemaps/routes.coffee#L42).
+
+```js
+SITEMAP_BASE_URL = 'http://artsy-sitemaps.s3-website-us-east-1.amazonaws.com'
+httpProxy = require 'http-proxy'
+{ parse } = require 'url'
+sitemapProxy = httpProxy.createProxyServer(target: SITEMAP_BASE_URL)
+
+@sitemaps = (req, res, next) ->
+  req.headers['host'] = parse(SITEMAP_BASE_URL).host
+  sitemapProxy.web req, res
+```
+
+### Results
+
+We schedule a workflow for each sitemap in Spark with Oozie. Each sitemap job completes in under 5 minutes, which is pretty remarkable given where we started. It will also easily scale to many multiples of our current size as our data continues to grow exponentially.
```

Comment:
`continues to grow exponentially` isn't quite accurate. I think `continues to grow` would be compelling enough.